File: tests/run_vectors.js
// Cross-language golden vector test runner (JavaScript side).
// Run: node tests/run_vectors.js
// Companion: tests/run_vectors.py must produce identical results against
// the same tests/test_vectors.json -- that agreement is the whole point.

const assert = require('assert');
const path = require('path');
const fs = require('fs');
const { processTelemetryPayload } = require('../telemetry_volume_engine.js');

const vectorsPath = path.join(__dirname, 'test_vectors.json');
const vectors = JSON.parse(fs.readFileSync(vectorsPath, 'utf8'));

let failures = 0;

for (const v of vectors) {
  const i = v.input;
  const actual = processTelemetryPayload(i.rawLaserDistanceMm, i.tLiquid, i.tLid, i.secondsDelayed);
  try {
    assert.deepStrictEqual(actual, v.expected);
  } catch (e) {
    failures += 1;
    console.log(`MISMATCH [${v.name}]`);
    console.log(`  expected: ${JSON.stringify(v.expected)}`);
    console.log(`  actual:   ${JSON.stringify(actual)}`);
  }
}

console.log(`\n${vectors.length - failures}/${vectors.length} vectors match`);
process.exit(failures === 0 ? 0 : 1);

File: tests/run_vectors.py
"""
Cross-language golden vector test runner (Python side).
Run: python tests/run_vectors.py
Companion: tests/run_vectors.js must produce identical results against the
same tests/test_vectors.json -- that agreement is the whole point.
"""

import json
import os
import sys

sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))
from telemetry_volume_engine import process_telemetry_payload

VECTORS_PATH = os.path.join(os.path.dirname(__file__), "test_vectors.json")


def main():
    with open(VECTORS_PATH) as f:
        vectors = json.load(f)

    failures = 0
    for v in vectors:
        i = v["input"]
        raw = i.get("rawLaserDistanceMm")
        t_liq = i.get("tLiquid", 4.0)
        t_lid = i.get("tLid", 45.0)
        sec = i.get("secondsDelayed", 4.0)
        actual = process_telemetry_payload(raw, t_liq, t_lid, sec)
        expected = v["expected"]
        if actual != expected:
            failures += 1
            print(f"MISMATCH [{v['name']}]")
            print(f"  expected: {expected}")
            print(f"  actual:   {actual}")

    print(f"\n{len(vectors) - failures}/{len(vectors)} vectors match")
    sys.exit(0 if failures == 0 else 1)


if __name__ == "__main__":
    main()
